### ISSUE TITLE
Octaves 1, 2, and 3 give parsing errors

### ISSUE THREAD

**Issue by jmmelko** [sky-music/sky-python-music-sheet-maker#78]
For most people, absolute octave numbers mean nothing.
If the song is within a single octave, they simply give note names (eg C, B, G) and that's fine because our code assumes the octave is 4.

But if the songs spans several octaves, some people will input the song starting from octave 1. This gives errors because we assume that Sky's piano starts at C4.

We have three options from here, I will list them by order of my preference:

1) Change 'note_shift' back to 'octave_shift', use 'octave_shift' *before* parsing the song to interpret user input
2) use note_shift *before* parsing the song. Then we do octave_shift = int(note_shift/n), and note_shift = note_shift - octave_shift*n, where n = 12 or 7, depending on which point of view we choose (real-world's chromatic or Sky's heptatonic system)
3) Try to parse the song with different octave_shift and chooses the one which gives less errors

My preferences goes to solution no1 because now that we handle all the musical keys available in Sky, we don't need to transpose songs within Sky's system (the initial use of note_shift was to transpose songs to work with C), while transposition within the general chromatic scale will be handled elsewhere (#32)

What do you think Tracey?

**Comment by t1-tracey** [comment]
Yeah I was thinking about this, out of the options I think 1 is best too. 

But I also think that maybe we could just set the default octave in the program to start from 1 (we are able to do this). When I type in songs, I never think of it in terms of the absolute octave either

**Comment by t1-tracey** [comment]
And then we can use the octave_shift for anyone who types in C4-C6 etc.

**Comment by jmmelko** [comment]
> And then we can use the octave_shift for anyone who types in C4-C6 etc.

**Comment by t1-tracey** [comment]
Wow nice, I will test it.

> I have kept the default octave to 4, because it is the one used by skysheetmusic.com and also in the middle of the piano keyboard.
> Also, if you start at 1 it becomes impossible to transpose, unless you use negative octave numbers.
> But if the majority convinces me that 1 is better, we'll make the change.

Alright I’m ok with that 👍

**Comment by t1-tracey** [comment]
Ooh I haven’t tested 1/7 etc. yet but the octave shift does work now, awesome 😊

**Comment by t1-tracey** [comment]
there is a merge conflict because the same lines of code were edited in both dev and beta branches. I am so annoyed arrrgh. 

I want all code to be committed to only the dev branch in future. Once a feature is complete in the dev branch, it will be merged into beta, and tested before being put into master.

**Comment by jmmelko** [comment]
Sorry, I thought at first that minor changes could be applied to beta directly! Shouldn’t happen again. Dev is a sandbox, beta is for testing, and stable is for sharing to everyone.

**Comment by t1-tracey** [comment]
Aw no worries 😊 Yeah you are confirming, I’m glad, that’s pretty much it 😊 I guess I expected dev is for all commits, beta is for a 'stable state' from dev (e.g. it might take a number commits before a feature such as western scales stops being 'in progress').

and master is for everyone to use (with archive being the backup, so that in case master breaks, there will be a backup to revert to) . But this can be open to revision, just going with this for now to reduce excess merge conflicts

**Comment by t1-tracey** [comment]
﻿Aw no worries 😊 Yeah you are confirming, I’m glad, that’s pretty much it 😊 I guess I expected dev is for all commits, beta is for a 'stable state' from dev (e.g. it might take a number commits before a feature such as western scales stops being 'in progress').

and master is for everyone to use (with archive being the backup, so that in case master breaks, there will be a backup to revert to) . But this can be open to revision, just going with this for now to reduce excess merge conflicts

**Comment by t1-tracey** [comment]
ok I did a test with a fraction of octaves, it seems the chosen song_key is not adjusted accordingly so I will try to fix that

**Comment by jmmelko** [comment]
I have analyzed the pitch of Sky instruments:

Piano C4
Harp C3
Bass harp C1
Guitar C3
Shamiukule C3
Blowing Horn C3
Pan flute C4
Flute C4

So we can stick to C4 because it is the one of the piano but there is pull towards C3!

**Comment by t1-tracey** [comment]
Awesome, I got the same ranges on the wiki too! I am impressed. i think this can work for midi output. 

in terms of typing in songs, I still think C D E F G A B C2 D2 E2 F2 G2 A2 B2 is more intuitive (and they are assumed as C4-C6) but I’m willing to go with C4-C6 and get some feedback on it.

**Comment by jmmelko** [comment]
> Awesome, I got the same ranges on the wiki too! I am impressed. i think this can work for midi output. in terms of typing in songs, I still think C D E F G A B C2 D2 E2 F2 G2 A2 B2 is more intuitive (and they are assumed as C4-C6) but I’m willing to go with C4-C6 and get some feedback on it.

You are almost right: it is indeed more intuitive provided that C is the lowest C in the song, which happens if C occurs early in the song, or if the player takes care to detect the lowest C first.
Otherwise, chances are that he will use C for the first C it finds, and then later the song gets down in pitch, and since C-1 does not exist, he will have to change its notation.
With an intermediate C4 it does not happen.
But I agree that even in this case if the user had to use a C3 he may be tempted to shift everything up so that the lowest C is C4, so the problem remains unchanged!

Mmmmhhhh... is it possible that we detect if the player uses notes with octave somewhere in the song (in find_key for instance) and then decide how to handle the song (relative or absolute scale)?

**Comment by jmmelko** [comment]
I think you’re right. We have, I think, 3 cases to handle:

1/ the player does not specify any octave at all: e.g. C D B E C B B F G.
It means that the song lies within 1 octave. Then we can assume that the default octave is 4, and C4 is (0,0) in Sky for the C key. Actually it doesn’t matter which default octave we choose internally, as long as it is mapped to (0,0) it’s fine.

2/ the player uses an undecorated note (e.g. ABCDEFG) for the starting octave, and a note with an octave number for the others. Most likely he will use C, C2, C3, so we can safely assume that 'C' == 'C1', and map C1 to (0,0). The case of a player using C for C4 is very unlikely.

3/ the player uses decorated notes, but sometimes forgets to add the octave, e.g.:
C4 E4 C5 B3 C D4 E4 D G G5

Then we have two solutions:
* automatically adapt the default octave to the middle octave of the song, by some intelligent coding in find_key(), and treat missing octaves as the middle octave
* do nothing, and treat missing octaves as the default octave, which would be 1. It is up to the player to use note_shift = -3 to parse his song correctly.

**Comment by t1-tracey** [comment]
Hi JMM, no worries :D I have just been trying to understand how to set up the discord bot, I have been able to do it successfully now!

Ah I see your point about not being able to go below that first C, if the player has accidentally treat it as the lowest C.

Hm I agree with the possible solutions that you've laid out, of handling the cases. I think at this stage, I want to get people to test the bot and see which cases across 1/ 2/ 3 are used most common, before making decisions on how smart we make the program

**Comment by jmmelko** [comment]
I don’t see where you guys are working on the bot’s code: the latest commit in that-sky-bot repository is dated 1 month ago, on the other hand you have been testing things on Discord’s forum (with the current code? But the bot’s version of main.py was clearly outdated)
I’d like to see the code to learn something in this domain.

**Comment by jmmelko** [comment]
If you make the change be careful to also change the numbers in the inverse_position_map in EnglishNoteParser and also in EnglishChords.

Maybe we can generate these dictionaries using the self.default_starting_octave property.
For instance:
self.inverse_position_map = {(0, 0): 'C'+str(self.default_starting_octave), (1,2):'C'+str(self.default_starting_octave+1)

or:
self.inverse_position_map = {(0, 0): 'C', (1,2):'C2'

Or:

default_octave_str=''
if self.default_starting_octave>1:
 default_octave_str=str(self.default_starting_octave)

self.inverse_position_map = {(0, 0): 'C'+default_octave_str, (1,2):'C'+str(self.default_starting_octave+1


I can make these boring changes after you have changed the default value of self.default_starting_octave and checked that it works.

Thank you.

**Comment by t1-tracey** [comment]
Oh yeah I changed `self.default_starting_octave` to 1 when I was testing the bot, and it worked. 

For the inverse_position_map, will it be replaced by the algorithm in #67 ? I can take care of that, just having trouble prioritising right now 😓

**Comment by jmmelko** [comment]
Thank you.
I will first do what I said in the message above.
As for the rest... I don’t think that’s urgent. Translation between notation formats is a bonus, not the core of the program.

**Comment by t1-tracey** [comment]
Ok 👍 I will focus on properly preparing our program for bot (in the bot-refactor branch, which I will then merge into dev)

**Comment by jmmelko** [comment]
Okay, I have made the fix in the dev branch. The EnglishChords dictionary is uglier than ever, but that was the price to pay. 
Please tell me if you want me to apply the fix to the bot branches directly. I will do nothing by default.

**Comment by t1-tracey** [comment]
it's ok it will be addressed in #67 :blush: I have merged it into bot-refactor and beta, I will have to test it later, but it looks good :+1:
